ibis/common/tests/test_patterns.py

```diff
@@ -401,24 +432,50 @@ def test_mapping_of():
     assert p.match({"foo": 1}, context={}) is NoMatch
 
 
-def test_object_pattern():
-    class Foo:
-        __match_args__ = ("a", "b")
+class Foo:
+    __match_args__ = ("a", "b")
 
-        def __init__(self, a, b):
-            self.a = a
-            self.b = b
+    def __init__(self, a, b):
+        self.a = a
+        self.b = b
 
-        def __eq__(self, other):
-            return type(self) == type(other) and self.a == other.a and self.b == other.b
+    def __eq__(self, other):
+        return type(self) == type(other) and self.a == other.a and self.b == other.b
+
+
+class Bar:
+    __match_args__ = ("c", "d")
+
+    def __init__(self, c, d):
+        self.c = c
+        self.d = d
+
+    def __eq__(self, other):
+        return type(self) == type(other) and self.c == other.c and self.d == other.d
 
+
+def test_object_pattern():
     p = Object(Foo, 1, b=2)
     o = Foo(1, 2)
     r = match(p, o)
     assert r is o
     assert r == Foo(1, 2)
 
 
+def test_object_pattern_complex_type():
+    p = Object(Not(Foo), 1, 2)
```

Comment:
This may be just my brain, but I would find the following easier to read:

```python
Not(Object(Foo, 1, 2))
```

Though now after staring at it for a minute I see that the pattern you've written has different semantics than what I wrote.

The one I wrote says: "match anything that isn't an instance of Foo constructed with arguments 1 and 2"

The one you wrote says: "match anything that was constructed with arguments 1 and 2, as long as it is **not** an instance of `Foo`"

Tricky!